Transparent fallback for BeautifulSoup lxml -> html5lib -> html.parser

pseyfert opened the pull request:
ran into problems getting down-frab-videos running. turned out the easiest fix is using html5lib as parser.

 * Added an error message giving the hint
 * Added a config switch

What do you think?

pseyfert committed:
add switch for lxml/html5lib replacement

mfherbst commented on down_frab_videos.py:
```diff
@@ -254,14 +279,15 @@ def get_format_list(media_prefix):
             'From': SOURCE,
         }
 
+        print("getting ", media_prefix + "/")
```
Any reason why you added this print statement?

pseyfert replied:
whoops. overlooked when cleaning up after debugging.

mfherbst replied:
Thought so ;).

mfherbst reviewed (commented):
Thanks for the PR.

It feels to me like a very hacky way to solve this. Is it not maybe better to use one of them by default and fall back to the other automatically if there are problems?

I mean I do not see the point, why this should be exposed to the config at all.

mfherbst commented:
**Update**: This comment is obsolete. Read on to the next ;)

~~Also I find it really funny you are getting this error. Because I never did on any of my systems. Any clue why?~~

~~Would it maybe make sense to have `html5lib` as the default and not have `lxml` at all? Would that solve your problem?~~

mfherbst commented:
Ok. I had a look at [the BeatifulSoup docs](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#installing-a-parser). From that I would suggest we have an automatic fallback mechanism `lxml`, `html5lib`, `html.parser`. If `lxml` cannot be found, I think we should issue a warning to suggest installation of the `lxml` python package.

What do you think?

mfherbst commented:
Maybe what I suggested above is a bit over the top after all. Does installing `lxml` solve your problem ... because if it does, than I would just bail out and issue an appropriate message, instructing to install the package.

pseyfert commented:
> It feels to me like a very hacky way to solve this.

Of cause it's hacky, I wanted to get the download started ;)

> I mean I do not see the point, why this should be exposed to the config at all.

Well, that's why I was asking for feedback. My thinking was that testing and
falling back once per video instead of once per installation doesn't seem
elegant. But either way is fine with me.

> If lxml cannot be found, I think we should issue a warning to suggest installation of the lxml python package.

That's the part I haven't understood so far about my failure, `lxml` is
installed, not sure why `BeautifulSoup` doesn't manage to pick it up. (I'm
undecided how dig I want to dig in understanding it.)

mfherbst commented:
> Of cause it's hacky, I wanted to get the download started ;)

I understand. Sorry for the inconvenience ;).

> My thinking was that testing and
falling back once per video instead of once per installation doesn't seem
elegant.

It does not happen once per video. BeautifulSoup is only needed at the initialisation stage for parsing the media files and the media formats, so I think having a built-in fallback, which is walked through each time is ok. And well ... due to the oddities and irregularities in the way the VOC does the language codes and so on the code is not really in the cleanest stage either.

> That's the part I haven't understood so far about my failure, lxml is
installed, not sure why BeautifulSoup doesn't manage to pick it up.

Strange. Well, I guess the automatic fall back works for both of us and we do not have to dig too much ;).

pseyfert committed:
rework html parser fallthrough

 - don't pick html parser from config, but use hard coded fallback
 - priorise lxml, html.parser, html5lib

mfherbst commented on down_frab_videos.py:
```diff
@@ -241,6 +243,28 @@ def __init__(self, short_message, long_message):
+        return BeautifulSoup(content, "html5lib")
+    except bs4.FeatureNotFound:
+        print("Error: could not parse with any of 'lxml', 'html.parser' and 'html5lib'. "
+              "Giving up.")
```
From the BeautifulSoup docs, I would say that `html.parser` is the worst and should be used last, right?

pseyfert replied:
Depends on your figure of merit. But let's put `html.parser` last.

mfherbst commented on down_frab_videos.py:
```diff
@@ -241,6 +243,28 @@ def __init__(self, short_message, long_message):
         self.long_message = long_message
 
 
+def wrap_bs4(content):
```
You could write this routine as a `for`-loop over a list `["lxml", "html.parser", "html5lib"]`. This makes it easer to extend the list at a later point.

pseyfert replied:
good point

mfherbst reviewed (changes requested):
Thanks for adapting. Apart from the order, just nits.

pseyfert committed:
review comments

mfherbst commented on down_frab_videos.py:
```diff
@@ -244,23 +244,18 @@ def __init__(self, short_message, long_message):
+            print("Warning: could not parse with {}, "
+                  "check your installation. ".format(parser))
+            if parser != known_parsers[-1]:
+                print("Falling back to next known parser.")
```
Would be nice to know which one it tried unsuccessfully.

pseyfert replied:
Maybe I don't understand, maybe you overlooked. That's the `{}` in line 255. For me it shows:

```
Warning: could not parse with lxml, check your installation. 
Falling back to next known parser.
```

mfherbst replied:
I overlooked.

mfherbst reviewed (approved):
Great work.
